plugin/src/main/java/org/opensearch/ml/sdkclient/XContentClient.java

```diff
@@ -0,0 +1,104 @@
+import org.opensearch.sdk.DeleteCustomResponse;
+import org.opensearch.sdk.GetCustomRequest;
+import org.opensearch.sdk.GetCustomResponse;
+import org.opensearch.sdk.PutCustomRequest;
+import org.opensearch.sdk.PutCustomResponse;
+import org.opensearch.sdk.SdkClient;
+
+public class XContentClient implements SdkClient {
+
+    private final Client client;
+    private final NamedXContentRegistry xContentRegistry;
+
+    public XContentClient(Client client, NamedXContentRegistry xContentRegistry) {
+        this.client = client;
+        this.xContentRegistry = xContentRegistry;
+    }
+
+    @Override
+    public CompletionStage<PutCustomResponse> putCustom(PutCustomRequest request) {
+        CompletableFuture<PutCustomResponse> future = new CompletableFuture<>();
+        try (XContentBuilder sourceBuilder = XContentFactory.jsonBuilder()) {
+            client
+                .index(
+                    new IndexRequest(request.index())
+                        .setRefreshPolicy(IMMEDIATE)
+                        .source(request.custom().toXContent(sourceBuilder, EMPTY_PARAMS)),
+                    ActionListener
+                        .wrap(
+                            r -> future.complete(new PutCustomResponse.Builder().id(r.getId()).created(r.getResult() == CREATED).build()),
+                            future::completeExceptionally
+                        )
+                );
+        } catch (IOException ioe) {
+            // Parsing error
+            future.completeExceptionally(ioe);
+        }
+        return future;
+    }
+
+    @Override
+    public CompletionStage<GetCustomResponse> getCustom(GetCustomRequest request) {
+        CompletableFuture<GetCustomResponse> future = new CompletableFuture<>();
+        client.get(new GetRequest(request.index(), request.id()), ActionListener.wrap(r -> {
+            try {
+                XContentParser parser = jsonXContent
+                    .createParser(xContentRegistry, LoggingDeprecationHandler.INSTANCE, r.getSourceAsString());
+                future
+                    .complete(
+                        new GetCustomResponse.Builder().id(r.getId()).custom(request.clazz().cast(new Object()).parse(parser)).build()
+                    );
+            } catch (Exception e) {
+                // Parsing error
+                future.completeExceptionally(e);
+            }
+        }, future::completeExceptionally));
+        return future;
+    }
+
+    @Override
+    public CompletionStage<DeleteCustomResponse> deleteCustom(DeleteCustomRequest request) {
+        CompletableFuture<DeleteCustomResponse> future = new CompletableFuture<>();
+        client
+            .delete(
+                new DeleteRequest(request.index(), request.id()),
+                ActionListener
+                    .wrap(
+                        r -> future.complete(new DeleteCustomResponse.Builder().id(r.getId()).deleted(r.getResult() == DELETED).build()),
```

Comment:
What happens if the deletion fails? How are we sending the error response back to the customer?

Reply:
The actionListener for the whole execution is being completed exceptionally.  I have not yet gone through to more carefully identify exception possibilities and give more detailed user information.  Again, this is an early-stage WIP.